dbms/src/DataTypes/DataTypesDecimal.h

```diff
@@ -320,12 +320,26 @@ inline std::enable_if_t<IsDataTypeNumber<FromDataType> && IsDataTypeDecimal<ToDa
 convertToDecimal(const typename FromDataType::FieldType & value, UInt32 scale)
 {
     using FromFieldType = typename FromDataType::FieldType;
+    using ToNativeType = typename ToDataType::FieldType::NativeType;
 
     if constexpr (std::is_floating_point_v<FromFieldType>)
     {
-        if (!std::isfinite(value))
-            throw Exception("Decimal convert overflow. Cannot convert infinity or NaN to decimal", ErrorCodes::DECIMAL_OVERFLOW);
-        return value * ToDataType::getScaleMultiplier(scale);
+        ToNativeType min_value;
+        ToNativeType max_value;
+        if constexpr (std::is_same_v<ToNativeType, Int128>)
+        {
+            min_value = __int128(0x8000000000000000ll) << 64; // min __int128
+            max_value = (__int128(0x7fffffffffffffffll) << 64) + 0xffffffffffffffffll; // max __int128
+        }
+        else
+        {
+            min_value = std::numeric_limits<ToNativeType>::min();
+            max_value = std::numeric_limits<ToNativeType>::max();
+        }
+        ToNativeType converted_value = static_cast<ToNativeType>(value * ToDataType::getScaleMultiplier(scale));
+        if (converted_value == min_value || converted_value == max_value)
```

Comment:
OK there's an error where we check isfinite(value) before multiplication and do not check the result. But why static_cast of INF returns min_value? Is it something documented in standard? If not we could just check the result.

```
auto out = value * ToDataType::getScaleMultiplier(scale);
if (!std::isfinite(value))
    throw Exception( ...
```

Reply:
This code works
```
    if constexpr (std::is_floating_point_v<FromFieldType>)
    {
        if (!std::isfinite(value))
            throw Exception("Decimal convert overflow. Cannot convert infinity or NaN to decimal", ErrorCodes::DECIMAL_OVERFLOW);

        auto out = value * ToDataType::getScaleMultiplier(scale);
        if constexpr (std::is_same_v<ToNativeType, Int128>)
        {
            static constexpr __int128 min_int128 = __int128(0x8000000000000000ll) << 64;
            static constexpr __int128 max_int128 = (__int128(0x7fffffffffffffffll) << 64) + 0xffffffffffffffffll;
            if (out < min_int128 || out > max_int128)
                throw Exception("Decimal convert overflow. Float is out of Decimal range", ErrorCodes::DECIMAL_OVERFLOW);
        }
        else
        {
            if (out < std::numeric_limits<ToNativeType>::min() || out > std::numeric_limits<ToNativeType>::max())
                throw Exception("Decimal convert overflow. Float is out of Decimal range", ErrorCodes::DECIMAL_OVERFLOW);
        }
        return out;
    }
```